pkg/server/singleprocess/service_job.go

```diff
@@ -603,17 +605,53 @@ func (s *Service) GetJobStream(
 				}
 			}
 
-			// If we haven't initialized output streaming and the output buffer
-			// is now non-nil, initialize that. This will send any buffered
-			// data down.
-			if eventsCh == nil && job.OutputBuffer != nil {
-				eventsCh, err = s.getJobStreamOutputInit(ctx, job, server)
-				if err != nil {
-					return err
+			if eventsCh == nil {
+				switch job.State {
+				case pb.Job_RUNNING:
+
+					// We're seeing the job start up live, so we'll initialize the
+					// event channel and use the job streamer to stream the logs
+					// in as they arrive via the runner interface.
+					// If the job OutputBuffer is nil, the streamer won't send
+					// any events.
+					eventsCh, err = s.getJobStreamOutputInit(ctx, log, job, server)
+					if err != nil {
+						msg := "failed to init job output stream"
+						log.Error(msg, "job.id", job.Id, "error", err)
+						return status.Error(codes.Internal, msg)
+					}
+
+				// NOTE: at present (2022-02-25) there is no exposed CLI or UI API that
+				// causes GetJobStream to be called on a completed job. Thusly this code below
+				// is entirely optimistic about future API usage. But it is tested!
+				case pb.Job_SUCCESS, pb.Job_ERROR:
+					// This means that the requested stream finished before GetJobStream was
+					// called. As such, we'll replay the output events.
+					events, err := s.logStreamProvider.ReadCompleted(ctx, log, s.state(ctx), job)
```

Comment:
This is the biggest shift. We didn't have a concept of explicitly reading completed jobs before - we would always read any buffered logs, then read streamed logs, and this case was handled by the buffered log reading.

Other server impls, though, benefit from getting an explicit trigger to read completed job logs - they can use it to pull logs out of persistent storage. Our implementation of ReadComplete here just reads buffered logs from memory.

Reply:
Totally makes sense